### ISSUE TITLE
Integration of C++ Algorithms (SVFit and FastMtt) into ColumnFlow

### ISSUE THREAD

**Issue by oponcet** [columnflow/columnflow#389]
Dear ColumnFlow experts,

Following your recent clarification on the framework, we have now reached the stage where we aim to incorporate more complex methods into our analysis. Specifically, we are interested in implementing SVFit and FastMtt, algorithms used to reconstruct the ditau mass using kinematic information from tau lepton decays and MET. This algorithm was previously implemented in C++ with ROOT usage in our last analysis ([IPHCProductionTools ClassicSVfit](https://github.com/alebihan/IPHCProductionTools/tree/master/TauAnalysis/ClassicSVfit)). Our goal now is to integrate it into ColumnFlow. There are two potential approaches: rewriting it entirely in Python with Awkward Array format to adapt it to the formalism of ColumnFlow, or using a wrapper to interface with the existing C++ classes developed for traditional event loop based method. We have several questions regarding this process:

- What would you recommend as the better approach: using a wrapper or rewriting the method in ColumnFlow? It's worth noting that the SVFit() function takes only `measured Taus`, `measuredMETx`, `measuredMETy`, and `covMET` variables as input and does not operate on arrays.
- If a wrapper is the preferred choice (for this case or another), which one do you recommend for ColumnFlow?
- If we decide to implement SVfit (or any other general method creating variables and associated columns), where would you suggest including it in the ColumnFlow structure? We are considering creating a `utils/` folder in the `analysis/` directory or in the `production/` folder. Are there any best practices you recommend for this situation?

Thank you for your guidance.

Océane

@gsaha009 @cardinia

**Comment by riga** [comment]
Hi Océane,

thanks for opening this discussion! I think this is very important to have as it will apply to many other algorithms for which a C(++) implementation already exists.

We haven't run into this case ourselves so far (in fact, we replaced SVFit in our analysis with a regression NN), but as you already mentioned there are two paths that one could follow and the following is just my personal view on the topic:

- If possible, I would rewrite the algorithm with scipy / numpy / awkward arrays (or even TensorFlow *). This also depends on whether the algorithm itself has other dependencies besides ROOT, but after a quick glimpse at the code you linked, this does not seem to be the case. However, I believe SVFit requires doing an actual fit / minimization which is not easy to model, but I know from other groups that it is definitely possible, even over multiple events at a time. For the latter, I know a group that did this with TensorFlow in the past (for the example for a single neutrino pt_Z reconstruction using a W mass constraint). I will ask them if they can share their code.

- If you succeed with porting SVFit, this should perhaps live in a dedicated repository, so that other groups that use different frameworks could also profit from it. A dedicated lightweight producer that simply invokes your algorithm could live in your analysis, or pushed *upstream* to cf if it's general enough.

- If it's not possible to port it, there is indeed the option to use a library such as [pybind11](https://pybind11.readthedocs.io/en/stable) (which is anyway available in most envs / sandboxes) to wrap it and call it within python. And it also seems like it should be able to [vectorize the wrapped function call](https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html#vectorizing-functions) (I'm not entirely sure though if this method simply provides function signature that allows array inputs, or if it actually "vectorizes" through parallel invocation, but I would suspect the former).

**Comment by riga** [comment]
Sounds reasonable. I guess I would have also started with the pybind11 approach.

On the integration of the software, unlike the SVFit code base, pybind11 should be provided through a sandbox. For that, have a look at how we do this in our analysis: https://github.com/uhh-cms/hh2bbtautau/tree/master/sandboxes
In your `<ANALYSIS_DIR>/sandboxes` directory, you only have to create a txt and a sh file that define the sandbox, with the txt file listing all packages you need (with `pybind11` being one of them).

You will likely create a producer to invoke wrapped SVFit algorithm, and this producer then just needs to declare that it needs your newly created sandbox to run. Here is an example for that: https://github.com/uhh-cms/hh2bbtautau/blob/master/hbt/production/hhbtag.py#L31

Hope this helps :)

**Comment by oponcet** [comment]
Thank you for the response. I also needed to create a ROOT object, so I followed the instructions to include ROOT:

I added `root~=6.00` to the file `sandboxes/root_sand.txt`.
I created a new .sh file `sandboxes/env_test.sh` which specifies the above-mentioned text file.
I added the line `sandbox=dev_sandbox("bash::$HCP_BASE/sandboxes/env_test.sh")` in my producer script.
Then, I attempted to import ROOT with `root = maybe_import("root")`.
However, it seems that it's not working. The ROOT module is mocked, and I suspect that ROOT is not found. I'm unsure if this is a sandbox issue or another problem. Bring to import pybind11 with the same procedure, I had a similar issue.

I also attempted to use `root = ROOT_import()` without success.

Additionally, I have a question regarding the compilation of the SVFit (C++) code. I'm wondering if I need to run CMake in my SVFit folder or if this is also something that I should link to the columnflow environment in some way.

Thank you

**Comment by riga** [comment]
Hi @oponcet ,

to get ROOT to work properly in a sandbox, I would recommend using a CMSSW-based sandbox as it is not so straight forward to compile ROOT from scratch based on a python version of a virtual env. You can find an example of a CMSSW sandbox (with the option to install custom python packages on top) here: https://github.com/columnflow/columnflow/blob/master/sandboxes/cmssw_columnar.sh
This would also solve the problem of compiling SVFit which can be done in the same step. I know a couple instances where SVFit is compiled in a default CMSSW env, and you can do the same here (and I'm sure you already have experience with that :) ).

For the use within python / cf, I can only refer you to the pybind11 documentation, as there is no mechanism (needed) in cf link anything. Our intention was to fully rely on tools like pybind11 to interface to C++ dependencies.

Hope this helps!

**Comment by oponcet** [comment]
Hi @riga ,

Thank you for your response. Your solution for importing CMSSW with the necessary package is indeed what we need. However, I am encountering several issues and I try several things. 

1. I attempted to utilize `sandbox=dev_sandbox("bash::$HCP_BASE/sandboxes/venv_columnar_tf.sh")` in my producer and import `tensorflow` as in your example. However, when printing `tf`, I received a `mock module`. When executing `law run cf.SelectEvents`, which calls the producer, I am unsure if I should receive the message `entering sandbox: venv_columnar_tf.sh,` but this message does not appear. It seems like the sandbox is not being used. Did I overlook something? I tried `CF_REINSTALL_SOFTWARE=1 source setup.sh`, but it did not resolve the issue.

2. Regarding the CMSSW environment, should I utilize the sandbox in the same manner? Should I use `sandbox=dev_sandbox("bash::$HCP_BASE/sandboxes/venv_cmssw_sand.sh")`, or is there another approach?

3. To avoid dealing with that, I attempted to change the default sandbox to `default_columnar_sandbox: bash::$CF_BASE/sandboxes/venv_cmssw_sand.sh` in `law.cfg`, but encountered the following error message: `/analysis/data/software/cmssw/venv_cmssw_sand_9f1ee8f3/CMSSW_12_6_2/venvs/columnar/bin/activate: No such file or directory. It appears I am missing an installation step here`. Could you please clarify how to properly use CMSSW environment ?

Thank you very much.

**Comment by pkausw** [comment]
Hi @oponcet , just a quick comment from the side lines: As of last Friday, we updated various sections of the official docs (available for the `latest` version, `stable` to follow soon), which now also contains a dedicated [section for sandboxes](https://columnflow.readthedocs.io/en/latest/user_guide/sandbox.html). This might help you to figure this out.

If you want to check if a given sandbox is installed properly, you could try the following command (e.g. if you want to check your venv `venv_columnar_tf` in the `dev` version)

```bash
cf_sandbox venv_columnar_tf_dev
```

This should give a bash terminal in this sandbox, in which you should also have e.g. tensorflow. This might help you debug a little more.

Finally: To reproduce your problem, it might be helpful if you could share your software version with us (e.g. commit hash of cf). 

Cheers,
Philip

**Comment by oponcet** [comment]
Hi @pkausw,

Thank you for your response. I've gone through the documentation, and it's really helpful! I appreciate your effort in pushing it forward.

I've been trying to follow the procedure outlined in the README, but it seems like it's not working out of the box as I expected. I believe I might be missing a step in the installation of the development sandbox. Let me walk you through my process, and perhaps you can identify what I'm missing.

1. For each new session, I execute:
   ```
   source setup.sh enveos
   ```

2. My default sandbox is defined here: [link to the sandbox configuration](https://github.com/oponcet/HCPToTauTau/blob/2d42a8994cd6395ac665b2e0295737b0f1831dc6/law.cfg#L39), and it is indeed the one being used when I run the code.

3. To build my development environment, I've created:
   - a `.sh` file in `/hcp/sandbox`: [link to the shell script](https://github.com/oponcet/HCPToTauTau/blob/SVFIT_dev/sandboxes/venv_columnar_tf.sh) which refers to
   - a `.txt` file in `/hcp/sandbox`: [link to the text file](https://github.com/oponcet/HCPToTauTau/blob/SVFIT_dev/sandboxes/columnar_tf.txt)

4. I specify the sandbox in one of my selectors: [link to the selector](https://github.com/oponcet/HCPToTauTau/blob/SVFIT_dev/hcp/selection/lepton.py#L37)

5. In the same file, I import TensorFlow using `maybe_import()`: [link to the import statement](https://github.com/oponcet/HCPToTauTau/blob/2d42a8994cd6395ac665b2e0295737b0f1831dc6/hcp/selection/lepton.py#L53). I also print `tf`.

6. Then, I run `law run cf.SelectEvents --version v1 --branch 1`, but:
   - I don't see the message `entering sandbox: venv_columnar_tf`; it's still running on my default sandbox.
   - The output of the print(`tf`) shows a mock module, indicating that the development environment is not being used.

I suspect I'm missing a step where the new environment should be installed. However, based on my understanding of the code and the README, it should happen because of step 4.

I did try `cf_sandbox venv_columnar_tf_dev`, and it installed the sandbox without any issue. I guess this is only for debuting and I should not add this line to the code. 

I hope this clarifies my issue. Thank you very much for your patience!

**Comment by pkausw** [comment]
Hi @oponcet ,

thanks for the links and your walk-through. I went through it and didn't find any obvious mistake in your configuration. In my opinion, this is also shown by the `cf_sandbox` test that you did - btw, this is a tool meant for debugging and should not be used in code since it returns an interactive shell to you.

I also saw that you use the latest version of cf, so it shouldn't be some legacy issue. Just to make sure: Are you sure that also the other submodules in your local setup are up to date? Git doesn't do so automatically, but you can make sure that everything is in the correct version by running `git submodule update --recursive` in your analysis repository.

If this doesn't help, I'll try to set up your analysis and try to reproduce this behavior locally. Sorry for the problems you have to experience!

**Comment by oponcet** [comment]
Hi @pkausw,

Thanks for the quick answer. I run `git submodule update --recursive`and it's seems up to date. Here are the versions that I'm using:
```
git submodule status --recursive
 d364181b08a61eecc27898226ab5d7aaeb6dd81a modules/cmsdb (heads/master)
 de11ccfe104a2e6b233153a568d8e4c1bf7b38be modules/columnflow (v0.2.2-403-gde11ccf)
 3f8b5c8954bfb1db7ba1c3bcf0265bc0f12eaa04 modules/columnflow/modules/law (v0.1.17-12-g3f8b5c8)
 12eedf2d5d72afc85c780e638275e81036df4efe modules/columnflow/modules/order (v0.1.10-197-g12eedf2)
 8b03ffa7c06cd9c8a38297b1c8923695d1ff1b07 modules/extern/pybind11 (v2.11.1-62-g8b03ffa7)
 ```

**Comment by oponcet** [comment]
Hi again,

Since I couldn't manage to run with the dev sandbox, I'm trying to use the CMSSW sandbox as the default sandbox in [law.cfg](https://github.com/oponcet/HCPToTauTau/blob/2d42a8994cd6395ac665b2e0295737b0f1831dc6/law.cfg#L40).

The sandbox is installing CMSSW_12_6_2, but I encountered the following error message: 

`Exception: trying to access 'gfal2.creat_context', but gfal2 is not installed`

It seems that I'm missing this package but I don't know exactly how to fix this. So I have the following questions:


1) Is this possible to set the `sandboxes/venv_cmssw_sand.sh` as the default sandbox? 

2) Do you have a similar file that include the same environment as 'sandboxes/venv_columnar.sh' plus the CMSSW environment?  Or how can I had the list of package needed for columnflow? 

3) If I need to add ROOT or SVFit package, where should I configure this in the file? 

Thanks for helping me with these basic questions.

**Comment by pkausw** [comment]
Hi @oponcet ,

I finally had chance to look at this a little closer, sorry for the delay. I think the communication regarding sandboxes this far has been a little misleading. 

The sandbox that is used/required by a given task (e.g. `SelectEvents`) is provided by the TaskArrayFunction instance that is _executed at command line level_. In you setup, this would mean that the sandbox for the `SelectEvents` task is defined by the Selector `main`, which in turn uses the `venv_columnar.sh`. So even though your `lepton` Selector defines its own sandbox, this information is not used since it's not the TaskArrayFunction instance that is actually called within the `SelectEvents` task. This has multiple reasons, mainly to facilitate a modular treatment of the sandboxes for e.g. remote jobs. Consequently, the top-level TaskArrayFunction instance (in your case your `main` Selector) needs to define a software environment that provides all packages that are required by the modules that are called during the call of this instance.

Regarding the follow up questions you've had:

- In order to use a `dev` environment, you also need to create a corresponding .sh script. You can use the `example.sh` and `example_dev.sh` scripts that are located in your sandbox directory (from the example analysis).
- In principle, the sandbox files for bash are "just" shell scripts. So whenever you need to do specific steps or load packages, you can write out exactly what needs to be done (note that you should perform some check whether the packages you'd like to load are already available, otherwise you potentionally do redundant steps). So in reference to a question you've had before, you could e.g. call the `make` command for SVFit directly within your shell script (if that is necessary).

Regarding the cmssw sand box issue: It's possible that CMSSW overwrite the default gfal version to be used. I've seen the issue you mention above on lxplus < 9 - is it possible that you have a similar problem?

**Comment by oponcet** [comment]
Hi,

Thank you for your response!

Regarding the CMSSW sandbox, I'm currently running it on lxplus8. I attempted to reinstall the environment on lxplus9 using `CF_REINSTALL_SOFTWARE=1 source setup.sh enveos`, and then proceeded to run the `sandboxes/venv_cmssw_sand.sh` as the default sandbox. However, I encountered the following error message: [output.log](https://github.com/columnflow/columnflow/files/14722466/output.log)

**Comment by oponcet** [comment]
Hi again,

I solved one issue with the CMSSW sandbox. CMSSW 12 was not compatible with lxplus9. Switching to CMSSW 13 solved this problem. However, I still encounter the 
```
error message: Exception: trying to access 'gfal2.creat_context', but gfal2 is not installed.
```

I noticed that one of the [files](https://github.com/riga/law/blob/d585cf8183c3e7b389ef86ed0dc4b2367d7eaae8/law/contrib/cms/scripts/setup_gfal_plugins.sh) in law mentions this problem; maybe it can help. It seems that the gfal and CMSSW incompatibility is already a known issue. If the problem persists, perhaps I could raise an issue on CMS Talk.

**Comment by oponcet** [comment]
Yes I can run on the default columnflow sandbox after installing CMSSW sandbox. But the problem is that if I want to use ROOT with my selector for instance, @riga advice me to run with CMSSW sandbox. Or maybe I didn't get your question?

**Comment by pkausw** [comment]
I think that heavily depends. I'm not an expert in what law provides you with exactly, but depending on what exactly you need from root, you might
- want to use python alternatives like `uproot` or something similar
- You can use code snippets like this
```python
foo = law.contrib.load("root")
ROOT = foo.import_ROOT()
```

Though I guess your requirement for ROOT originally comes from SVFit rather than anything else right? Sorry, I'm also not overly familiar with using anything ROOT-based within a law setup. I just thought that if you can install ROOT directly via pip, you might get around having to use the cmssw sandbox in the first place.

**Comment by oponcet** [comment]
Yes, that's what I'm thinking as well. And it seems that SVFit can be used without CMSSW as well. So we can avoid this problem. If you could provide an example of the `dev_dandbox.sh` file that I can use to install ROOT, it would be nice because I couldn't manage when I tried...

**Comment by pkausw** [comment]
Hi @oponcet ,

Please find examples for the files you'll need to write below. Some thoughts from my side:
- note that in principle, you don't *have* to use a `dev` sandbox. It's just meant to provide you with additional debugging tools you won't need for a large-scale processing of your data
- You will probably need to try out how to properly intergrate root into your python environment. The suggestion here is based on [this](https://stackoverflow.com/questions/66418716/pyroot-in-virtualenv) thread, but if that doesn't work you might want to try the import via law I mentioned in a post further above (or something else entirely)
- also, note that we are currently working on the procedure for the software setup, so while this answer should work right now, it might change a little in the future.

Let me know how it goes!

Cheers,
Philip


First, write a requirement file that can be parsed by `pip`. In this example, I directly included `pybind11` - I had a brief look at their docs, and don't think you actually _need_ to add it as a submodule to your repo. Nevertheless, I'm not an expert, so if you do need it as a submodule, just remove that line. Note that the first line in this file must be a version tag.

<details><summar>${HCP_BASE}/sandboxes/requirements_svfit.txt</summary> 

```
# version 1
pybind11  # would be better to actually pin this to a version
pyroot      # same here
```

</details>

Now you can create a shell script for each environment you want to create. In your case, you want to create to environments:
- one for production, i.e. without additional debugging tools
- one for local development, which also includes debugging tools 

<details><summar>${HCP_BASE}/sandboxes/venv_svfit_integration.sh</summary> 

```bash
#!/usr/bin/env bash

# Script that sets up a virtual env in $CF_VENV_BASE.
# For more info on functionality and parameters, see the generic setup script _setup_venv.sh.

action() {
    local shell_is_zsh="$( [ -z "${ZSH_VERSION}" ] && echo "false" || echo "true" )"
    local this_file="$( ${shell_is_zsh} && echo "${(%):-%x}" || echo "${BASH_SOURCE[0]}" )"
    local this_dir="$( cd "$( dirname "${this_file}" )" && pwd )"

    # set variables and source the generic venv setup
    export CF_SANDBOX_FILE="${CF_SANDBOX_FILE:-${this_file}}"
    export CF_VENV_NAME="$( basename "${this_file%.sh}" )"
    # put our new requirements here
    export CF_VENV_REQUIREMENTS="${HCP_BASE}/requirements_svfit.txt"

    source "${CF_BASE}/_setup_venv.sh" "$@"
}
action "$@"
```

</details>

For the dev version, you simply add more requirements, e.g. the ones that we use throughout columnflow

<details><summar>${HCP_BASE}/sandboxes/venv_svfit_integration_dev.sh</summary> 

```bash
#!/usr/bin/env bash

# Script that sets up a virtual env in $CF_VENV_BASE.
# For more info on functionality and parameters, see the generic setup script _setup_venv.sh.

action() {
    local shell_is_zsh="$( [ -z "${ZSH_VERSION}" ] && echo "false" || echo "true" )"
    local this_file="$( ${shell_is_zsh} && echo "${(%):-%x}" || echo "${BASH_SOURCE[0]}" )"
    local this_dir="$( cd "$( dirname "${this_file}" )" && pwd )"

    # set variables and source the generic venv setup
    export CF_SANDBOX_FILE="${CF_SANDBOX_FILE:-${this_file}}"
    export CF_VENV_NAME="$( basename "${this_file%.sh}" )"
    # put our new requirements here, this time also with the dev requirements
    export CF_VENV_REQUIREMENTS="${HCP_BASE}/requirements_svfit.txt,${CF_BASE}/dev.txt"

    source "${CF_BASE}/_setup_venv.sh" "$@"
}
action "$@"
```

</details>

**Comment by oponcet** [comment]
Hi,

Thank you for your response. I may have misunderstood, but to my knowledge, PyROOT cannot be installed via pip. I've only seen instructions for installing it with conda, and it's why I'm struggling with it and why I cannot use `requirement.txt` file.

Cheers,

Océane

**Comment by pkausw** [comment]
Ah, OK! In principle you should have access to your micromamba setup though, right? We don't manage the venvs with micromamba, so the structure is
- the micromamba base as, well, the base
- the difference python venvs, which all have the micromamba base as part of their path.

So in principle, after you have sourced your analysis-level `setup.sh` setup, you should be able to run a micromamba installation command. Of course, if you change your setup and would have to reinstall micromamba again, you'd have to do it again. At the moment the micromamba packages are hidden deeply within the setup.sh file in columnflow, but this might change in the future.

So long story short, if you install something in micromamba, all your venvs (both existing and new) should have access to it I think. Doing this in a more automated way would be nice, but right now installing additional micromamba packages once should be a solution to this problem.

**Comment by pkausw** [comment]
Hi @oponcet , 

are there any updates on this issue? Were you able to resolve it?

If yes, we would appreciate it if you could close this issue :)

**Comment by oponcet** [comment]
Hi @pkausw, @riga 

I've successfully resolved all my issues and implemented SVFit with the pybind11 wrapper. Here's a summary of what I tried, what didn't work, and how I resolved it:

- Regarding the problem with importing ROOT, I attempted to add it to the micromamba installation [here](https://github.com/columnflow/columnflow/blob/de11ccfe104a2e6b233153a568d8e4c1bf7b38be/setup.sh#L604), but encountered issues with Python 3.6 compatibility with ROOT and missing dependencies, particularly libcppyy_backend3_9.so, which I found to be a known issue [here](https://root-forum.cern.ch/t/cannot-import-root-6-22-in-python-3-9-1-could-not-load-cppyy-backend-library/43764/2).

To bypass this issue, I opted to use a local version of ROOT (in my case, it was already installed in lxplus). Before each session and configuring my ColumnFlow setup, I imported the necessary libraries and sourced ROOT with:

```
LD_LIBRARY_PATH=/usr/local/lib
CPPYY_BACKEND_LIBRARY=/cvmfs/sft.cern.ch/lcg/app/releases/ROOT/6.30.04/x86_64-centosstream9-gcc113-opt/lib/libcppyy_backend3_9.so
source /cvmfs/sft.cern.ch/lcg/app/releases/ROOT/6.30.04/x86_64-centosstream9-gcc113-opt/bin/thisroot.sh 
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/cvmfs/sft.cern.ch/lcg/app/releases/ROOT/6.30.04/x86_64-centosstream9-gcc113-opt/lib/
```
Then, it was possible to import ROOT like: `root = maybe_import("ROOT")`

- For pybind11, I simply had to import the correct module produced by the pybind11 wrapper. At the beginning of my file, I imported my module as shown in my [svfit.py](https://github.com/oponcet/HCPToTauTau/blob/b09537f1be2063f8b9bf38df16db29a8d1643f42/hcp/production/svfit.py#L21C1-L22C1). Similarly to ROOT, I ensured that I imported the correct library in `$LD_LIBRARY_PATH`.

- Another issue, not related to ColumnFlow but worth mentioning since it's related to the wrapper, was that ROOT is wrapped in Python with cppyy. I encountered some trouble using Python ROOT objects as arguments for wrapped C++ functions that required C++ ROOT objects. I had to adapt my wrapper to handle this situation. Hence, it might be easier to use the cppyy wrapper in these kinds of situations instead of pybind11.
You can find the SVFit code in ColumnFlow [here](https://github.com/oponcet/HCPToTauTau/blob/SVFIT_dev/hcp/production/svfit.py) and the wrapper [here](https://github.com/oponcet/ClassicSVfit/blob/fastMTT_19_02_2019/wrapper/pybind_wrapper.cpp).

Now everything is running smoothly, and in the end, I'm not using a custom sandbox or cmssw.

Thanks again for all the answers you provided, for your time and your help ! :)

**Comment by pkausw** [comment]
Thanks for the final summary of your solution @oponcet , happy all is working for you now :)
